dev/test/pool.ts

```diff
@@ -327,4 +327,25 @@ describe('Client pool', () => {
         expect(err).to.equal('The client has already been terminated');
       });
   });
+
+  it('waits for existing operations to complete before releasing clients', async () => {
+    const clientPool = new ClientPool<{}>(1, 0, () => {
+      return {};
+    });
+    const deferred = new Deferred<void>();
+
+    const op = clientPool.run(REQUEST_TAG, async () => {
+      await deferred.promise;
+      return Promise.resolve('success');
+    });
+    const terminate = clientPool
+      .terminate()
+      .then(() => Promise.resolve('success'));
```

Comment:
You don't need to return a resolved Promise here, but you could use the `then` callback to modify some global state. That allows you to verify that it hasn't run:

```
let terminated = false;

const terminateOp = clientPool.terminate(() => { terminate = true });

// Give the Promise chain a chance to resolve

setImmediate(() => {
   expect(terminate).to.be.false;
   
   await terminateOp;
   expect(terminated).to.be.true;
   done(); // New done() callback
});

```

Reply:
This is very clean, thanks!